discovery-handlers/onvif/src/discovery_handler.rs

```diff
@@ -72,42 +72,48 @@ impl DiscoveryHandler for DiscoveryHandlerImpl {
         tokio::spawn(async move {
+            let mut prev_cameras = Vec::new();
+            let mut filtered_camera_devices = HashMap::new();
             loop {
+                let mut changed_camera_list = false;
                 let onvif_query = OnvifQueryImpl {};
 
                 trace!("discover - filters:{:?}", &discovery_handler_config,);
-                let discovered_onvif_cameras = util::simple_onvif_discover(Duration::from_secs(
-                    discovery_handler_config.discovery_timeout_seconds as u64,
-                ))
-                .await
-                .unwrap();
-                trace!("discover - discovered:{:?}", &discovered_onvif_cameras,);
-                // apply_filters never returns an error -- safe to unwrap
-                let filtered_onvif_cameras = apply_filters(
-                    &discovery_handler_config,
-                    discovered_onvif_cameras,
-                    &onvif_query,
+                let mut socket = util::get_socket().await.unwrap();
+                let latest_cameras = util::simple_onvif_discover(
+                    &mut socket,
+                    discovery_handler_config.scopes.as_ref(),
+                    Duration::from_secs(discovery_handler_config.discovery_timeout_seconds as u64),
                 )
                 .await
                 .unwrap();
-                trace!("discover - filtered:{:?}", &filtered_onvif_cameras);
-                let mut changed_camera_list = false;
-                let mut matching_camera_count = 0;
-                filtered_onvif_cameras.iter().for_each(|camera| {
-                    if !cameras.contains(camera) {
+                trace!("discover - discovered:{:?}", &latest_cameras);
+                // Remove cameras that have gone offline
+                prev_cameras.iter().for_each(|c| {
+                    if !latest_cameras.contains(c) {
                         changed_camera_list = true;
-                    } else {
-                        matching_camera_count += 1;
+                        filtered_camera_devices.remove(c);
                     }
                 });
-                if changed_camera_list || matching_camera_count != cameras.len() {
-                    trace!("discover - sending updated device list");
-                    cameras = filtered_onvif_cameras.clone();
+
+                // Insert newly discovered camera that are not filtered out
+                for c in &latest_cameras {
+                    if !prev_cameras.contains(c) {
+                        if let Some(device) =
+                            apply_filters(&discovery_handler_config, &c, &onvif_query).await
+                        {
+                            changed_camera_list = true;
+                            filtered_camera_devices.insert(c.clone(), device);
+                        }
+                    }
+                }
+                prev_cameras = latest_cameras;
```

Comment:
nit: can `prev_cameras = latest_cameras` be moved into the `if changed_camera_list` scope?